### PR TITLE
Per-Object Motion Blur

### LINKED ISSUE #4710: Request add Soft shadow and motion blur function
Request add  Soft shadow and motion blur function。

### PR DESCRIPTION
**Pull request by aevyrie**
https://github.com/bevyengine/bevy/assets/2632925/e046205e-3317-47c3-9959-fc94c529f7e0

# Objective

- Adds per-object motion blur to the core 3d pipeline. This is a common effect used in games and other simulations.
- Partially resolves #4710

## Solution

- This is a post-process effect that uses the depth and motion vector buffers to estimate per-object motion blur. The implementation is combined from knowledge from multiple papers and articles. The approach itself, and the shader are quite simple. Most of the effort was in wiring up the bevy rendering plumbing, and properly specializing for HDR and MSAA.
- To work with MSAA, the MULTISAMPLED_SHADING wgpu capability is required. I've extracted this code from #9000. This is because the prepass buffers are multisampled, and require accessing with `textureLoad` as opposed to the widely compatible `textureSample`.
- Added an example to demonstrate the effect of motion blur parameters.

## Future Improvements

- While this approach does have limitations, it's one of the most commonly used, and is much better than camera motion blur, which does not consider object velocity. For example, this implementation allows a dolly to track an object, and that object will remain unblurred while the background is blurred. The biggest issue with this implementation is that blur is constrained to the boundaries of objects which results in hard edges. There are solutions to this by either dilating the object or the motion vector buffer, or by taking a different approach such as https://casual-effects.com/research/McGuire2012Blur/index.html
- I'm using a noise PRNG function to jitter samples. This could be replaced with a blue noise texture lookup or similar, however after playing with the parameters, it gives quite nice results with 4 samples, and is significantly better than the artifacts generated when not jittering.

---

## Changelog

- Added: per-object motion blur. This can be enabled and configured by adding the `MotionBlurBundle` to a camera entity.

### CODE REVIEW

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+@group(0) @binding(0) var screen_texture: texture_2d<f32>;
+@group(0) @binding(1) var motion_vectors: texture_2d<f32>;
+@group(0) @binding(2) var depth: texture_depth_2d;
+#endif
+@group(0) @binding(3) var texture_sampler: sampler;
+struct PostProcessSettings {
+    shutter_angle: f32,
+    max_samples: u32,
+    depth_bias: f32,
+#ifdef SIXTEEN_BYTE_ALIGNMENT
```
Comment (aevyrie) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:24]: Is this needed? Due to needing to sample the prepass, I think this can only run on native and webgpu.

**Comment by aevyrie** [comment]
CI is failing on check bans for some reason. 🤷

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 2]
```diff
@@ -0,0 +1,83 @@
+#import bevy_pbr::prepass_utils
+#import bevy_pbr::utils
+#import bevy_core_pipeline::fullscreen_vertex_shader FullscreenVertexOutput
+
+#ifdef MULTISAMPLED
```
Comment (aevyrie) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:5]: Multisampling specialization certainly makes things a bit more complex. I think this is a necessary evil though.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        weight_total = 1.0;
+    }
+
+    return accumulator / weight_total;
+}
+
+fn noise(frag_coord: vec2<f32>) -> f32 {
+    let k1 = vec2<f32>(23.14069263277926, 2.665144142690225);
+    return fract(cos(dot(frag_coord, k1)) * 12345.6789);
+}
```
Comment (DGriffin91) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:83]: This should use blue noise. It also should change from frame to frame so TAA resolves it more nicely and weird temporal artifacts don't show up.

If you're going to use white noise, at least use an implementation with a better distribution like:
```wgsl
fn uhash(a: u32, b: u32) -> u32 { 
    var x = ((a * 1597334673u) ^ (b * 3812015801u));
    // from https://nullprogram.com/blog/2018/07/31/
    x = x ^ (x >> 16u);
    x = x * 0x7feb352du;
    x = x ^ (x >> 15u);
    x = x * 0x846ca68bu;
    x = x ^ (x >> 16u);
    return x;
}

fn unormf(n: u32) -> f32 { 
    return f32(n) * (1.0 / f32(0xffffffffu)); 
}

fn hash_noise(ifrag_coord: vec2<i32>, frame: u32) -> f32 {
    let urnd = uhash(u32(ifrag_coord.x), (u32(ifrag_coord.y) << 11u) + frame);
    return unormf(urnd);
}
```

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    let shutter_angle = settings.shutter_angle;
+    let texture_size = vec2<f32>(textureDimensions(screen_texture));
+    let frag_coords = vec2<i32>(in.uv * texture_size);
+
+    let motion_vector = textureLoad(motion_vectors, frag_coords, i32(sample_index)).rg;
+    let exposure_vector = shutter_angle * motion_vector;
+    let speed = length(exposure_vector * texture_size);
+    let n_samples = i32(clamp(speed * 2.0, 1.0, f32(settings.max_samples)));
+
+    let this_depth = textureLoad(depth, frag_coords, i32(sample_index));
```
Comment (DGriffin91) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:45]: I think a webgl2 could be supported if it omitted the depth test.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 3 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:83]: Is there a uniform that tracks frame number somewhere? I did a search and could not find one.

Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:83]: Ah, globals.frame_count

**Comment by JMS55** [comment]
Needs rebasing, please

File: crates/bevy_core_pipeline/src/motion_blur/mod.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+        let Ok(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        render_app
+            .init_resource::<SpecializedRenderPipelines<pipeline::MotionBlurPipeline>>()
+            .add_systems(
+                Render,
+                (pipeline::prepare_motion_blur_pipelines.in_set(RenderSet::Prepare),),
```
Comment (torsteingrindvik) [inline crates/bevy_core_pipeline/src/motion_blur/mod.rs:85]:
```suggestion
                pipeline::prepare_motion_blur_pipelines.in_set(RenderSet::Prepare),
```

File: crates/bevy_core_pipeline/src/motion_blur/mod.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    /// The shutter angle describes the fraction of a frame that a camera's shutter is open and
+    /// exposing the film/sensor. For 24fps cinematic film, a shutter angle of 0.5 (180 degrees) is
+    /// common. This means that the shutter was open for half of the frame, or 1/48th of a second.
+    /// The lower the shutter angle, the less exposure time and thus less blur. A value greater than
+    /// one is unrealistic and results in an object's blur stretching further than it traveled in
+    /// that frame.
+    pub shutter_angle: f32,
+    /// The upper limit for how many samples will be taken per-pixel. The number of samples taken
+    /// depends on the speed of an object.
+    pub max_samples: u32,
```
Comment (torsteingrindvik) [inline crates/bevy_core_pipeline/src/motion_blur/mod.rs:41]: I noticed the shader returns base color if <2 max samples, so perhaps mention an ok range for users?

**Comment by aevyrie** [comment]
I cannot replicate the CI failure. The output log looks suspiciously bugged as well. Not sure what is going on there.

**Comment by JMS55** [comment]
@aevyrie the associated type name has changed, replace ViewQuery with ViewData.

File: crates/bevy_core_pipeline/src/motion_blur/mod.rs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+};
+
+pub mod node;
+pub mod pipeline;
+
+/// Adds [`MotionBlur`] and the required depth and motion vector prepasses to a camera entity.
+#[derive(Bundle, Default)]
+pub struct MotionBlurBundle {
+    pub motion_blur: MotionBlur,
+    pub depth_prepass: crate::prepass::DepthPrepass,
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/mod.rs:25]: Nit: Import prepass types

File: crates/bevy_core_pipeline/src/motion_blur/mod.rs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+            MOTION_BLUR_SHADER_HANDLE,
+            "motion_blur.wgsl",
+            Shader::from_wgsl
+        );
+        app.add_plugins((
+            ExtractComponentPlugin::<MotionBlur>::default(),
+            UniformComponentPlugin::<MotionBlur>::default(),
+        ));
+
+        let Ok(render_app) = app.get_sub_app_mut(RenderApp) else {
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/mod.rs:126]: The entire render app setup should be moved to Plugin::finish()

File: crates/bevy_core_pipeline/src/motion_blur/mod.rs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    fn finish(&self, app: &mut App) {
+        let Ok(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        let render_device = render_app.world.resource::<RenderDevice>().clone();
+
+        render_app.insert_resource(pipeline::MotionBlurPipeline::new(&render_device));
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/mod.rs:159]: Rather than do this, implement FromWorld for your pipeline and use init_resource()

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    let texture_size = vec2<f32>(textureDimensions(screen_texture));
+    let frag_coords = vec2<i32>(in.uv * texture_size);
+
+#ifdef MULTISAMPLED
+    let base_color = textureLoad(screen_texture, frag_coords, i32(sample_index));
+#else
+    let base_color = textureSample(screen_texture, texture_sampler, in.uv);
+#endif
+
+    if i32(settings.samples) <= 0 || settings.shutter_angle <= 0.0 {
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:43]: Can you just not run the node instead? (skip creating the pipeline)

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    // This means that for a frame time of 20ms, the shutter is only open for 10ms.
+    //
+    // Using a shutter angle larger than 1.0 is non-physical, objects would need to move further
+    // than they physically travelled during a frame, which is not possible.
+    let exposure_vector = shutter_angle * this_motion_vector;
+
+    var accumulator: vec4<f32>;
+    var weight_total = 0.0;
+    let n_samples = i32(settings.samples);
+    let noise = hash_noise(frag_coords, globals.frame_count); // 0 to 1
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:79]: If this is meant to be used with TAA, consider using interleaved gradient noise.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+    #ifdef MULTISAMPLED
+        let sample_color = textureLoad(screen_texture, sample_coords, i32(sample_index));
+    #else
+        let sample_color = textureSample(screen_texture, texture_sampler, sample_uv);
+    #endif
+    #ifdef MULTISAMPLED
+        let sample_motion = textureLoad(motion_vectors, sample_coords, i32(sample_index)).rg;
+    #else
+        let sample_motion = textureSample(motion_vectors, texture_sampler, sample_uv).rg;
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:95]: Do you need to check that motion vectors don't point off screen?

File: crates/bevy_core_pipeline/src/motion_blur/node.rs [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        let post_process = view_target.post_process_write();
+
+        let msaa = world.resource::<Msaa>();
+        let layout = if msaa.samples() == 1 {
+            &motion_blur_pipeline.layout
+        } else {
+            &motion_blur_pipeline.layout_msaa
+        };
+
+        let bind_group = render_context.render_device().create_bind_group(
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/node.rs:65]: Use the new bind group API please.

File: crates/bevy_core_pipeline/src/motion_blur/pipeline.rs [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+    pub(crate) sampler: Sampler,
+    pub(crate) layout: BindGroupLayout,
+    pub(crate) layout_msaa: BindGroupLayout,
+}
+
+impl MotionBlurPipeline {
+    pub(crate) fn new(render_device: &RenderDevice) -> Self {
+        let mb_layout = &[
+            // The screen texture
+            BindGroupLayoutEntry {
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/pipeline.rs:36]: Use the newer bind group layout builder API please

File: crates/bevy_core_pipeline/src/motion_blur/pipeline.rs [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+        let mut shader_defs = vec![];
+
+        match key.samples {
+            1 => (),
+            _ => shader_defs.push(ShaderDefVal::from("MULTISAMPLED")),
+        }
+
+        #[cfg(all(feature = "webgl", target_arch = "wasm32"))]
+        {
+            shader_defs.push("NO_ARRAY_TEXTURES_SUPPORT".into());
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/pipeline.rs:201]: Why did you add this shader def?

File: crates/bevy_core_pipeline/src/motion_blur/pipeline.rs [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+                shader_defs,
+                entry_point: "fragment".into(),
+                targets: vec![Some(ColorTargetState {
+                    format: if key.hdr {
+                        ViewTarget::TEXTURE_FORMAT_HDR
+                    } else {
+                        TextureFormat::bevy_default()
+                    },
+                    // blend: Some(BlendState::REPLACE),
+                    blend: None,
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/pipeline.rs:221]: Why is this None?

File: crates/bevy_pbr/src/render/mesh_view_bindings.rs [hunk 17]
```diff
@@ -297,7 +297,12 @@ pub fn generate_view_layouts(
         #[cfg(debug_assertions)]
         let texture_count: usize = entries
             .iter()
-            .filter(|entry| matches!(entry.ty, BindingType::Texture { .. }))
+            .filter(|entry| {
```
Comment (JMS55) [inline crates/bevy_pbr/src/render/mesh_view_bindings.rs:300]: Why was this changed?

**Review by JMS55 — CHANGES_REQUESTED** [review]
I think it would improve the example to make the camera track the object by default. Also, does TAA improve this effect? If so, please document that.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 11 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:79]: It can be used with or without TAA. I'm not familiar with interleaved gradient noise or how it would interact with various AA types. This noise was added based on a previous review of this PR.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 10 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:43]: Can you do that on a per-camera basis? I thought that was global.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 11 above)
Comment (DGriffin91) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:79]: iirc motion blur is supposed to run after TAA

Comment (DGriffin91) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:79]: It would be nice to use blue noise here. This is decent white noise, but there will probably be quite a bit less variance with blue noise.

File: crates/bevy_core_pipeline/src/motion_blur/mod.rs (hunk 7 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/mod.rs]: I don't know what you are asking for here. As far as I can tell this matches all the other render bundles.

File: crates/bevy_core_pipeline/src/motion_blur/node.rs (hunk 13 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/node.rs:65]: Can you provide docs or something? I have no idea what that is.

File: crates/bevy_core_pipeline/src/motion_blur/pipeline.rs (hunk 16 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/pipeline.rs:220]: Because that's what all the other similar pipelines do.

File: crates/bevy_core_pipeline/src/motion_blur/node.rs (hunk 13 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/node.rs:65]: The docs on `BindGroup` point me to `create_bind_group`, which Im already using.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 11 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:80]: Blue noise is out of scope for this PR. The implementation is sufficient as-is.
This runs after TAA.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 12 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:96]: no, they are only used to check motion similarity

File: crates/bevy_core_pipeline/src/motion_blur/node.rs (hunk 13 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/node.rs:65]: Updated main and copied one of the other pipelines

File: crates/bevy_core_pipeline/src/motion_blur/mod.rs (hunk 7 above)
Comment (JMS55) [reply crates/bevy_core_pipeline/src/motion_blur/mod.rs]: Instead of doing crate:prepass::Foo, just import Foo directly.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 10 above)
Comment (JMS55) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:43]: You should be using ViewNode, which will run the node per-camera, and you can specify that you require MotionBlurSettings or whatever component in the view query.

File: crates/bevy_core_pipeline/src/motion_blur/pipeline.rs (hunk 16 above)
Comment (DGriffin91) [reply crates/bevy_core_pipeline/src/motion_blur/pipeline.rs:137]: None is faster

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 10 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:43]: > you can specify that you require MotionBlurSettings or whatever component in the view query.

Will that solve the problem here, where I want to skip the node if the samples or shutter angle are <= 0?

Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl]: I think I figured this out in [8be1d96](https://github.com/bevyengine/bevy/pull/9924/commits/8be1d96f5627281b6bb7f61559e31e05a46dfa9e)

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+        let sample_depth = textureSample(depth, texture_sampler, sample_uv);
+    #endif
+    #endif
+
+        var weight = 1.0;
+        let is_sample_in_fg = !(depth_supported && sample_depth < this_depth && sample_depth > 0.0);
+        if is_sample_in_fg {
+            // The following weight calculation is used to eliminate ghosting artifacts that are
+            // common in motion-vector-based motion blur implementations. While some resources
+            // recommend using depth, I've found that sampling the velocity results in significantly
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:109]: Instead of depth, you can try plane distance, which has less artifacts on parallel surfaces. May or may no work, I'm not going to block this PR on it by any means, but if you're interested: https://developer.download.nvidia.com/video/gputechconf/gtc/2020/presentations/s22699-fast-denoising-with-self-stabilizing-recurrent-blurs.pdf#page=44

Also, did you make sure to linearize the depth before doing depth comparisons?

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    #ifdef MULTISAMPLED
+        let sample_depth = textureLoad(depth, sample_coords, i32(sample_index));
+    #else
+        let sample_depth = textureSample(depth, texture_sampler, sample_uv);
+    #endif
+    #endif
+
+        var weight = 1.0;
+        let is_sample_in_fg = !(depth_supported && sample_depth < this_depth && sample_depth > 0.0);
+        if is_sample_in_fg {
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:106]: Rather than using an if statement, you can also turn it into a binary 0/1 weight that you multiply the result by to avoid branching. Depends on how coherent you think threadgroups will be.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+            // sample was occluding the fragment?"
+            //
+            // Note to future maintainers: proceed with caution when making any changes here, and
+            // ensure you check all oclusion/disocclusion scenarios and fullscreen camera rotation
+            // blur for regressions.
+            let frag_speed = length(step_vector);
+            let sample_speed = length(sample_motion) / 2.0; // Halved because the sample is centered
+            let cos_angle = dot(step_vector, sample_motion) / (frag_speed * sample_speed * 2.0);
+            let motion_similarity = clamp(abs(cos_angle), 0.0, 1.0);
+            if sample_speed * motion_similarity < frag_speed {
```
Comment (JMS55) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:129]: Same thing here, turn an if statement into a weight using f32(condition) and multiply to avoid the branch.

**Review by JMS55 — APPROVED** [review]
Shader could avoid some branches, use a better noise (IGN/STBN), and could experiment with some established surface comparison heuristics (plane distance, normals) (commonly used in denoising), but it works well enough and I don't want to block on further, possibly futile, experimentation. I'm happy merging it as-is.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 19 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:106]: I don't think that can make a difference either way here, there is no else clause, so no matter what, worst case scenario is that the thread has to compute what is inside the block, which is equivalent to dissolving the block, computing it every time, and multiplying by a weight.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 18 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:109]: I don't understand the connection here, can you add some detail? The depth test is only used to test for occlusion (in front or behind?), I don't see how linearization or the plane distance matters here, I just need a boolean pass/fail.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 20 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:129]: Is that actually going to make any difference? This should optimize away completely. I'd think we would want to prioritize readability in cases like this where there should be no perf impact.

**Comment by aevyrie** [comment]
Can anyone recommend a way to profile performance? I'm dubious to make perf changes without being able to meaningfully measure perf.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 20 above)
Comment (DGriffin91) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:129]: I don't think this `if` will be a branch. If we're really worried we could use a select. (or check the code gen) I think the cast to f32 and multiply would definitely be more ops.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 18 above)
Comment (JMS55) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:109]: You mentioned there was difficulties accidentally sampling a static foreground object when trying to blur the background. You can compare the depth of each sample to the depth of the pixel you're trying to blur, to have a heuristic to detect different surfaces. It's commonly used to prevent blurring and temporal ghosting across different surfaces in denoisers.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 19 above)
Comment (JMS55) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:106]: Makes sense

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 20 above)
Comment (JMS55) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:129]: Makes sense

**Comment by JMS55** [comment]
Make sure you have a reproducible exact frame setup, and then run it through Nvidia's NSight GPU Frame Profiler, or AMD's equivalent (Radeon Graphics Analyzer iirc). You'll get per-pass timings, as well as statistics on shader occupancy, per-line stalls and timings, etc.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 18 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:109]: > You can compare the depth of each sample to the depth of the pixel you're trying to blur, to have a heuristic to detect different surfaces

Yeah, this is what I spent a ton of time on, and found that depth heuristics are always fragile, because that isn't a good way to seperate objects for blur, but motion similarity is.

**Comment by JMS55** [comment]
I really wouldn't worry about the perf, lets just leave it as-is.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl (hunk 18 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:109]: The problem I found with depth heuristics is that for MB, the most annoying artifact you want to reduce is background bleeding. What you are trying to do with the depth heuristic is figure out if the thing you are sampling could've affected the current pixel, using a depth test is an indirect way to answer the question by answering a _different_ question, which is "is the thing I'm sampling the same object as the thing I'm writing to". Instead, you can answer the _actual_ question by looking at the motion vectors, because it tells you directly whether or not that sample could've been located at the current pixel, based on the speed and direction it is moving.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+            // object to blur when it should be sharp ("background bleeding"). This is most obvious
+            // when the camera is tracking a fast moving object. The tracked object should be sharp,
+            // and should not bleed into the motion blurred background.
+            //
+            // To attenuate these incorrect samples, we compare the motion of the fragment being
+            // blurred to the UV being sampled, to answer the question "is it possible that this
+            // sample was occluding the fragment?"
+            //
+            // Note to future maintainers: proceed with caution when making any changes here, and
+            // ensure you check all oclusion/disocclusion scenarios and fullscreen camera rotation
```
Comment (pcwalton) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:123]: typo: "oclusion" -> "occlusion"

File: crates/bevy_core_pipeline/src/motion_blur/pipeline.rs [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+            1 => vec![self.layout.clone()],
+            _ => vec![self.layout_msaa.clone()],
+        };
+
+        let mut shader_defs = vec![];
+
+        match key.samples {
+            1 => (),
+            _ => shader_defs.push(ShaderDefVal::from("MULTISAMPLED")),
+        }
```
Comment (pcwalton) [inline crates/bevy_core_pipeline/src/motion_blur/pipeline.rs:115]: Why not just `if key.samples > 1`?

File: examples/3d/motion_blur.rs [hunk 23]
```diff
@@ -0,0 +1,401 @@
...
+                            .with_rotation(Quat::from_rotation_z(std::f32::consts::FRAC_PI_2)),
+                        ..default()
+                    },
+                    Rotates,
+                ));
+            };
+            spawn_wheel(1.0, 1.0);
+            spawn_wheel(1.0, -1.0);
+            spawn_wheel(-1.0, 1.0);
+            spawn_wheel(-1.0, -1.0);
```
Comment (pcwalton) [inline examples/3d/motion_blur.rs:176]: Cute :)

File: examples/3d/motion_blur.rs [hunk 24]
```diff
@@ -0,0 +1,401 @@
...
+
+// Parametric function for a looping race track. `offset` will return the point offset normal to the center of the race course.
+fn race_track_shape(offset: f32, t: f32) -> Vec2 {
+    let x_tweak = 2.0;
+    let y_tweak = 3.0;
+    let scale = 8.0;
+    let x0 = (x_tweak * t).sin();
+    let y0 = (y_tweak * t).cos();
+    let dx = x_tweak * (x_tweak * t).cos();
+    let dy = y_tweak * -(y_tweak * t).sin();
```
Comment (pcwalton) [inline examples/3d/motion_blur.rs:305]: I think this would be clearer as:

```rust
let dt = vec2(x_tweak * (x_tweak * t).cos(), y_tweak * (-y_tweak * t).sin()).normalize_or_zero();
vec2(x0 + offset, y0 - offset) * dt.yx() * scale
```

Up to you.

File: examples/3d/motion_blur.rs [hunk 25]
```diff
@@ -0,0 +1,401 @@
...
+        },
+        bevy::render::render_resource::TextureDimension::D2,
+        &texture_data,
+        bevy::render::render_resource::TextureFormat::Rgba8UnormSrgb,
+        bevy::render::render_asset::RenderAssetUsages::RENDER_WORLD,
+    );
+    img.sampler = ImageSampler::Descriptor(ImageSamplerDescriptor {
+        address_mode_u: ImageAddressMode::Repeat,
+        address_mode_v: ImageAddressMode::MirrorRepeat,
+        mag_filter: bevy::render::texture::ImageFilterMode::Nearest,
```
Comment (pcwalton) [inline examples/3d/motion_blur.rs:397]: nit: Is there a reason why you imported `ImageSampler` and `ImageAddressMode` but not `ImageFilterMode`/`TextureDimension`/`Extent3d`/`TextureFormat`/`RenderAssetUsages`?

File: examples/3d/motion_blur.rs [hunk 26]
```diff
@@ -0,0 +1,401 @@
...
+    let mut palette = [
+        164, 164, 164, 255, 168, 168, 168, 255, 153, 153, 153, 255, 139, 139, 139, 255, 153, 153,
+        153, 255, 177, 177, 177, 255, 159, 159, 159, 255,
+    ];
+
+    let mut texture_data = [0; TEXTURE_SIZE * TEXTURE_SIZE * 4];
+    for y in 0..TEXTURE_SIZE {
+        let offset = TEXTURE_SIZE * y * 4;
+        texture_data[offset..(offset + TEXTURE_SIZE * 4)].copy_from_slice(&palette);
+        palette.rotate_right(12);
```
Comment (pcwalton) [inline examples/3d/motion_blur.rs:380]: TIL this method exists.

**Review by pcwalton — COMMENTED** [review]
I didn't look at this too closely but it looks fine to me, none of these nits should be considered blocking.

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+                // cover enough distance to reach the original frag, there is no way it could have
+                // influenced this frag at all, and should be discarded.
+                weight = 0.0;
+            }
+        }
+        weight_total += weight;
+        accumulator += weight * sample_color;
+    }
+
+    let has_moved_less_than_a_pixel = length(this_motion_vector * texture_size) < 1.0;
```
Comment (pcwalton) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:140]: slightly faster: `dot(this_motion_vector * texture_size, this_motion_vector * texture_size) < 1.0`

File: crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+fn uhash(a: u32, b: u32) -> u32 { 
+    var x = ((a * 1597334673u) ^ (b * 3812015801u));
+    // from https://nullprogram.com/blog/2018/07/31/
+    x = x ^ (x >> 16u);
+    x = x * 0x7feb352du;
+    x = x ^ (x >> 15u);
+    x = x * 0x846ca68bu;
+    x = x ^ (x >> 16u);
+    return x;
+}
```
Comment (pcwalton) [inline crates/bevy_core_pipeline/src/motion_blur/motion_blur.wgsl:164]: I think these should be moved to `utils.wgsl`, as that's where our other rand-like functions live.

**Review by pcwalton — APPROVED** [review]
This looks good to me, just a couple of nits.

File: examples/3d/motion_blur.rs (hunk 25 above)
Comment (aevyrie) [reply examples/3d/motion_blur.rs:397]: Ah, I could move the imports to the top of the function. Should clean that up.

File: crates/bevy_core_pipeline/src/motion_blur/pipeline.rs (hunk 22 above)
Comment (aevyrie) [reply crates/bevy_core_pipeline/src/motion_blur/pipeline.rs:115]: copy-pasta and a bunch of merges with main, I assume. :)

File: examples/3d/motion_blur.rs (hunk 25 above)
Comment (JMS55) [reply examples/3d/motion_blur.rs:397]: I would just use glob imports for wgpu stuff, if you have more than ~7.

File: examples/3d/motion_blur.rs (hunk 24 above)
Comment (aevyrie) [reply examples/3d/motion_blur.rs:305]: This ended up breaking things, so I kept it unchagned.

**Comment by alice-i-cecile** [comment]
I'm comfortable with the code quality here, and this is a useful standard rendering feature. Merging :)

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1305 if you'd like to help out.
